plugins/sparc2_alignment_data_collector.py

```diff
@@ -0,0 +1,726 @@
+        self.ax.set_ylabel("Mirror s (m)", fontsize=10)
+        self.ax.set_zlabel("Stage z (m)", fontsize=10)
+        self.ax.set_title("Acquisition Progress")
+        self.ax.scatter(l, s, z, c="gray", alpha=0.5)
+        self.ax.scatter([l[-1]], [s[-1]], [z[-1]], c="red", label="Current")
+        self.ax.scatter(
+            self._aligned_pos["l"],
+            self._aligned_pos["s"],
+            self._aligned_pos["z"],
+            c="blue",
+            label="Aligned",
+        )
+        self.ax.legend()
+        self.canvas.draw()
+
+    def _run_acquisition(
+        self,
+        minus_dl: float,
+        plus_dl: float,
+        minus_ds: float,
+        plus_ds: float,
+        minus_dz: float,
+        plus_dz: float,
+        nl: int,
+        ns: int,
+        nz: int,
+        path: str
+    ):
+        """
+        Perform the mirror alignment cube acquisition.
+
+        Iterates over the specified ranges for mirror (l, s) and stage (z) positions,
+        acquires images at each position, updates metadata, saves the data asynchronously,
+        and updates the progress plot and status. Handles cancellation and pausing requests,
+        and restores the original aligned position when finished or paused.
+
+        :param minus_dl: Negative offset for mirror l axis.
+        :param plus_dl: Positive offset for mirror l axis.
+        :param minus_ds: Negative offset for mirror s axis.
+        :param plus_ds: Positive offset for mirror s axis.
+        :param minus_dz: Negative offset for stage z axis.
+        :param plus_dz: Positive offset for stage z axis.
+        :param nl: Number of steps for mirror l axis.
+        :param ns: Number of steps for mirror s axis.
+        :param nz: Number of steps for stage z axis.
+        :param path: Directory path to save acquired data.
+        """
+        l0 = self._aligned_pos["l"]
+        s0 = self._aligned_pos["s"]
+        z0 = self._aligned_pos["z"]
+        l_min, l_max = self.mirror.axes["l"].range
+        s_min, s_max = self.mirror.axes["s"].range
+        z_min, z_max = self.stage.axes["z"].range
+
+        l_values = numpy.clip(numpy.linspace(l0 - minus_dl, l0 + plus_dl, nl), l_min, l_max)
+        s_values = numpy.clip(numpy.linspace(s0 - minus_ds, s0 + plus_ds, ns), s_min, s_max)
+        z_values = numpy.clip(numpy.linspace(z0 - minus_dz, z0 + plus_dz, nz), z_min, z_max)
+
+        visited_l, visited_s, visited_z = [], [], []
+
+        try:
+            idx = 0
+            # Pause the ccd stream
+            self.tab._ccd_stream.is_active.value = False
+            self.tab._ccd_stream.should_update.value = False
+            # Pausing the ccd stream causes the e-beam to be blanked and mode as internal
+            # make the mode as external and unblank the beam to get valid ccd data
+            self.ebeam.external.value = True
+            self.ebeam.blanker.value = False
+            for l in l_values:
```

Comment:
@pieleric , @tepals  there is also a possibility to make this loop "Starts at the aligned center and moves outwards in expanding shells". This can help solve the drift issue over time in `l and s` and acquire good initial data at the aligned position first. Maybe this approach has its own issue and can take more time.

Reply:
Sounds like a good approach! It's probably a good idea to have at least one iteration of the prediction, so having more accurate data closer to alignment will improve the final iteration.
And +1 for the 3D hilbert curve :grinning: